File: Sources/CountriesFlagsObjC/source/Countries/UIImage+Niue.m
//
//  UIImage+Niue.m
//  CountriesFlags
//

#import "Countries/UIImage+Niue.h"

#import "Extensions/UIBezierPath+Star.h"
#import "Extensions/CGSize+Extension.h"

#import "Countries/UIImage+UnitedKingdom.h"

@implementation UIImage (Niue)

+ (UIImage * _Nullable)flagNiueWithSize:(CGSize)size {
    UIColor* colorYellow = [UIColor colorWithRed:254.0f/255.0f
                                           green:221.0f/255.0f
                                            blue:0.0f
                                           alpha:1.0f];
    UIColor* colorBlue = [UIColor colorWithRed:1.0f/255.0f
                                         green:33.0f/255.0f
                                          blue:105.0f/255.0f
                                         alpha:1.0f];
    
    UIGraphicsBeginImageContextWithOptions(size, NO, 0.0f);
    
    [colorYellow setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));

    [[UIImage flagUnitedKingdomWithSize:CGSizeHalfSize(size)] drawAtPoint:CGPointZero];
    
    [colorBlue setFill];
    [[UIBezierPath bezierPathWithArcCenter:CGPointMake(size.width / 4.0f, size.height / 4.0f)
                                    radius:size.height / 30.0f * 2.5f
                                startAngle:0.0f
                                  endAngle:M_PI * 2.0f
                                 clockwise:YES] fill];

    [colorYellow setFill];
    [[UIBezierPath starAtPoint:(CGPointMake(size.width / 9.6f, size.height / 3.91f))
                        radius:size.height / 30.0f * 1.5f
                          type:FlagStarTypeFivePointed] fill];
    [[UIBezierPath starAtPoint:(CGPointMake(size.width / 100.0f * 39.55f, size.height / 3.91f))
                        radius:size.height / 30.0f * 1.5f
                          type:FlagStarTypeFivePointed] fill];
    [[UIBezierPath starAtPoint:(CGPointMake(size.width / 4.0f, size.height / 11.45f))
                        radius:size.height / 30.0f * 1.5f
                          type:FlagStarTypeFivePointed] fill];
    [[UIBezierPath starAtPoint:(CGPointMake(size.width / 4.0f, size.height / 2.37f))
                        radius:size.height / 30.0f * 1.5f
                          type:FlagStarTypeFivePointed] fill];
    [[UIBezierPath starAtPoint:(CGPointMake(size.width / 4.0f, size.height / 4.0f))
                        radius:size.height / 30.0f * 2.5f
                          type:FlagStarTypeFivePointed] fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

@end

File: Sources/CountriesFlags/Countries/UIImage+Niue.swift
//
//  UIImage+Niue.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_Niue "wikipedia"
     
     Generates the flag of Niue.
     For more information see [wikipedia].
     
     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of Niue.
     */
    internal static func flagNiue(with size: CGSize) -> UIImage {
        let colorYellow = UIColor(red: 254.0 / 255.0,
                                  green: 221.0 / 255.0,
                                  blue: 0.0,
                                  alpha: 1.0)
        let colorBlue = UIColor(red: 1.0 / 255.0,
                                green: 33.0 / 255.0,
                                blue: 105.0 / 255.0,
                                alpha: 1.0)

        UIGraphicsBeginImageContextWithOptions(size, true, 0.0)

        colorYellow.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        flagUnitedKingdom(with: size.halfSize).draw(at: .zero)

        colorBlue.setFill()
        UIBezierPath(arcCenter: CGPoint(x: size.width / 4.0, y: size.height / 4.0),
                     radius: size.height / 30.0 * 2.5,
                     startAngle: 0.0,
                     endAngle: .pi * 2.0,
                     clockwise: true).fill()

        colorYellow.setFill()
        UIBezierPath.star(at: CGPoint(x: size.width / 9.6, y: size.height / 3.91),
                          radius: size.height / 30.0 * 1.5).fill()
        UIBezierPath.star(at: CGPoint(x: size.width / 100.0 * 39.55, y: size.height / 3.91),
                          radius: size.height / 30.0 * 1.5).fill()
        UIBezierPath.star(at: CGPoint(x: size.width / 4.0, y: size.height / 11.45),
                          radius: size.height / 30.0 * 1.5).fill()
        UIBezierPath.star(at: CGPoint(x: size.width / 4.0, y: size.height / 2.37),
                          radius: size.height / 30.0 * 1.5).fill()
        UIBezierPath.star(at: CGPoint(x: size.width / 4.0, y: size.height / 4.0),
                          radius: size.height / 30.0 * 2.5).fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image ?? UIImage()
    }
}

#endif
